server/samples/clients/node/appencryption_client.js

```diff
@@ -0,0 +1,274 @@
+                    self.#decryptResolve(decryptedPayload);
+                    break;
+                case 'error_response':
+                    logger.info('error received: ' + sessionResponse.error_response.message);
+                    break;
+                default:
+                    self.#getSessionResolve();
+            }
+        });
+
+        this.#call = call;
+    }
+
+    /**
+     * Create a new session for communicating with the gRPC server
+     * @param partition
+     * @returns {Promise<>}
+     */
+    getSession(partition) {
+        let self = this;
+        return new Promise((resolve, err) => {
+            self.#getSessionResolve = resolve;
+
+            self.#call.write({get_session: {partition_id: partition,},});
+        });
+    }
+
+    /**
+     * Encrypt data using the current session and its partition.
+     * @param payload
+     * @returns {Promise<>}
+     */
+    encrypt(payload) {
+        let self = this;
+        return new Promise((resolve, err) => {
+            self.#encryptResolve = resolve;
+            self.#call.write({encrypt: {data: Buffer.from(payload),},});
+        });
+    }
+
+    /**
+     * Decrypt the data using the current session and its partition
+     * @param drr
+     * @returns {Promise<>}
+     */
+    decrypt(drr) {
+        let self = this;
+        return new Promise(function (resolve, err) {
+            self.#decryptResolve = resolve;
+            self.#call.write({decrypt: {data_row_record: drr}});
+        });
+    }
+
+    /**
+     * Close the stream
+     */
+    close() {
+        this.#call.end()
+    }
+}
+
+/**
+ * Generate a random string of fixed length
+ * @param length
+ * @returns {string}
+ */
+function randomString(length = 12) {
+    let result = '';
+    let characters = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
+    let charactersLength = characters.length;
+    for (let i = 0; i < length; i++) {
+        result += characters.charAt(Math.floor(Math.random() * charactersLength));
+    }
+
+    return result;
+}
+
+/**
+ * Runs the client
+ * @param client
+ * @returns {Promise<void>}
+ */
+async function run_client(client) {
+    let payload = randomString();
+
+    logger.info(`encrypting payload ${payload}`);
+    let drr = await client.encrypt(payload);
+
+    logger.info(`calling decrypt`);
+    let decryptedPayload = await client.decrypt(drr);
+
+    if (decryptedPayload != payload) {
+        logger.info('oh no... something went terribly wrong!');
+    } else {
+        logger.info(`test completed successfully\n`);
+    }
+}
+
+/**
+ * Executes run_client once
+ * @param client
+ * @returns {Promise<void>}
+ */
+async function run_once(client) {
+
+    await run_client(client);
+    client.close();
+}
+
+/**
+ * Executes run_client until the process is interrupted.
+ * @param client
+ * @returns {Promise<void>}
+ */
+
+async function run_continuously(client) {
+    while (true) {
+        process.on('SIGINT', () => {
```

Comment:
Should handle SIGTERM here as well as it's the signal issued by docker on container stop

Reply:
The signal listeners to be moved outside of the loop. While it's currently functional you really only need one listener per signal for the process.